src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -3920,32 +3921,39 @@ extern "C" CLR_BOOL QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollid
     ExInfo* pExInfo = pThis->GetNextExInfo();
     bool isCollided = false;
 
-    MethodDesc *pMD = pThis->m_crawl.GetFunction();
+    bool doingFuncletUnwind = pThis->m_crawl.IsFunclet();
+    PCODE preUnwindControlPC = pThis->m_crawl.GetRegisterSet()->ControlPC;
+
+    retVal = pThis->Next();
+    if (retVal == SWA_FAILED)
+    {
+        EH_LOG((LL_INFO100, "SfiNext (pass=%d): failed to get next frame", pTopExInfo->m_passNumber));
+        goto Exit;
+    }
 
-    // Check for reverse pinvoke (but eliminate the case when the caller is managed) or CallDescrWorkerInternal.
-    if (!ExecutionManager::IsManagedCode(GetIP(pThis->m_crawl.GetRegisterSet()->pCallerContext)))
+    // Check for reverse pinvoke or CallDescrWorkerInternal.
+    if (pThis->GetFrameState() == StackFrameIterator::SFITER_NATIVE_MARKER_FRAME)
     {
-        CLR_BOOL invalidRevPInvoke;
+        EECodeInfo codeInfo(preUnwindControlPC);
```

Comment:
One last nit - I wonder if we should just make a copy of the other unwind info to avoid the lookup inside EECodeInfo initialization. We already have some cases of making a copy of that structure, e.g. in the StackFrameIterator.

Reply:
I considered that for a moment and there are some drawbacks.

On x86 we'd end up copying the whole structure with the GC info cache even if we don't need it. That reduces the effectiveness of the optimization from ~5.5% to ~2.2%. Still better than status quo but not ideal.

On non-x86 the `EECodeInfo` is small enough that it's probably possible to copy it with one or two SIMD instructions.

If you are fine with special casing x86 not to do it I can implement that.

Reply:
Let's keep it as is, since it is not on a hot path.